app/javascript/components/widget-controller.jsx

```diff
@@ -1,66 +1,85 @@
 import React from 'react';
+import PropTypes from 'prop-types';
+import gql from 'graphql-tag';
 import { mathMod } from 'ramda';
+import { Query } from 'react-apollo';
 import styled from 'styled-components';
 import FullPanel from '@components/full-panel';
 import SidePanel from '@components/side-panel';
+import { DisconnectedMessage, LoadingMessage } from '@messages/message';
 import Twitter from '@widgets/twitter';
 import Numbers from '@widgets/numbers';
 import Weather from '@widgets/weather';
 import Traffic from '@widgets/traffic';
 
-const SWITCH_INTERVAL = 20000;
+const getWidgets = gql`
+  query getWidgets($id: Int!) {
+    primaryLocation {
+      widgets {
+        enabled {
+          id
+          name
+          sidebarText
+        }
+        byIdOrFirst(id: $id) {
+          id
+          name
+          durationSeconds
+          sidebarText
+          showWeather
+        }
+      }
+    }
+  }
+`;
 
 const Wrapper = styled.div`
   outline: none;
 `;
 
-const widgets = [
-  {
-    panel: Weather.Panel,
-    text: 'Weather',
-    showWeather: false,
-  },
-  {
-    panel: Twitter.Panel,
-    text: '@MojoTech',
-    showWeather: true,
-  },
-  {
-    panel: Numbers.Panel,
-    text: 'MojoTech by the Numbers',
-    showWeather: true,
-  },
-  {
-    panel: Traffic.Panel,
-    text: 'Traffic',
-    showWeather: true,
-    hourStart: 16,
-    hourStop: 23,
-    isVisible: false,
-  },
-];
+const widgetElements = {
+  Weather,
+  Twitter,
+  Numbers,
+  Traffic,
+};
 
 export class WidgetController extends React.Component {
   constructor(props) {
     super(props);
     this.state = {
-      index: 0,
-      transitionTime: Date.now() + SWITCH_INTERVAL,
+      currentWidgetId: 0,
+      transitionTime: null,
     };
   }
 
   componentDidMount() {
+    const prefetchDelay = 1500;
+
+    this.lastPrefetchTime = 0;
     this.timerId = setInterval(() => {
       const { transitionTime } = this.state;
 
       if (!transitionTime) {
         return;
       }
 
+      const { client } = this.props;
+      const prefetchTime = transitionTime - prefetchDelay;
       const now = Date.now();
 
       if (transitionTime < now) {
-        this.moveDown();
+        this.nextWidget();
+      } else if (
+        prefetchTime < now &&
+        this.lastPrefetchTime < now - prefetchDelay
+      ) {
+        const { nextWidgetId } = this.state;
+        this.lastPrefetchTime = now;
+        client.query({
```

Comment:
do we have hooks here? hooks would be nice.
https://www.apollographql.com/docs/react/api/react-hooks/#uselazyquery

Reply:
I'm not sure what hooks would improve here.